File: 2016/9/python/main.py
import sys

def solve_part1(data):
    decompressed_length = 0
    i = 0
    n = len(data)
    while i < n:
        if data[i] == '(':
            close_paren_idx = data.find(')', i)
            if close_paren_idx == -1:
                raise ValueError(f"Unmatched parenthesis starting at index {i}")

            marker_content = data[i + 1:close_paren_idx]
            try:
                length, repetitions = map(int, marker_content.split('x'))
            except ValueError:
                raise ValueError(f"Invalid marker format: ({marker_content})")

            decompressed_length += length * repetitions
            i = close_paren_idx + 1 + length
        else:
            decompressed_length += 1
            i += 1
    return decompressed_length

def calculate_decompressed_length_recursive(data, start, end):
    decompressed_length = 0
    i = start
    while i < end:
        if data[i] == '(':
            close_paren_idx = data.find(')', i)
            if close_paren_idx == -1 or close_paren_idx >= end:
                 raise ValueError(f"Unmatched or out-of-bounds parenthesis starting at index {i}")

            marker_content = data[i + 1:close_paren_idx]
            try:
                length, repetitions = map(int, marker_content.split('x'))
            except ValueError:
                 raise ValueError(f"Invalid marker format: ({marker_content})")

            sub_start = close_paren_idx + 1
            sub_end = sub_start + length

            if sub_end > end:
                raise ValueError(f"Marker at {i} references data beyond segment end {end}")

            sub_length = calculate_decompressed_length_recursive(data, sub_start, sub_end)
            decompressed_length += sub_length * repetitions
            i = sub_end
        else:
            decompressed_length += 1
            i += 1
    return decompressed_length

def solve_part2(data):
    return calculate_decompressed_length_recursive(data, 0, len(data))

input_file_path = "../input.txt"

try:
    with open(input_file_path, 'r') as f:
        input_data = f.read().strip()
except FileNotFoundError:
    print(f"Error: {input_file_path} not found.")
    input_data = None
except Exception as e:
    print(f"Error reading file: {e}")
    input_data = None


if input_data:
    try:
        part1_length = solve_part1(input_data)
        print("="*20," Part 1", "="*20, f"\nDecompressed length = {part1_length}")
        part2_length = solve_part2(input_data)
        print("="*20," Part 2", "="*20, f"\nDecompressed length = {part2_length}")
    except ValueError as e:
        print(f"Processing Error: {e}")
    except RecursionError:
         print("Error: Maximum recursion depth exceeded. The input might be extremely deeply nested.")

File: 2016/9/rust/src/main.rs
use std::error::Error;
use std::fs;
use std::str;

// --- Part 1 ---
fn solve_part1(data: &[u8]) -> Result<u64, &'static str> {
    let mut decompressed_length: u64 = 0;
    let n = data.len();
    let mut i = 0;

    while i < n {
        if data[i] == b'(' {
            // Find the closing parenthesis relative to current position 'i'
            let close_paren_offset = data[i + 1..]
                .iter()
                .position(|&b| b == b')')
                .ok_or("Unmatched parenthesis")?;
            let close_paren_idx = i + 1 + close_paren_offset;

            let marker_content_bytes = &data[i + 1..close_paren_idx];
            let marker_content_str =
                str::from_utf8(marker_content_bytes).map_err(|_| "Invalid UTF-8 in marker")?;

            let (len_str, rep_str) = marker_content_str
                .split_once('x')
                .ok_or("Invalid marker format: missing 'x'")?;

            let length: usize = len_str.parse().map_err(|_| "Invalid length in marker")?;
            let repetitions: u64 = rep_str
                .parse() // Use u64 directly for repetitions
                .map_err(|_| "Invalid repetitions in marker")?;

            decompressed_length += (length as u64) * repetitions;

            // Move index past marker and the data block it refers to
            i = close_paren_idx + 1 + length;
            if i > n && length > 0 {
                // Ensure skip doesn't overshoot if length > 0
                return Err("Marker referred past end of data");
            }
        } else {
            decompressed_length += 1;
            i += 1;
        }
    }
    Ok(decompressed_length)
}

// --- Part 2 ---
fn calculate_decompressed_length_recursive(
    data: &[u8],
    start: usize,
    end: usize,
) -> Result<u64, &'static str> {
    let mut decompressed_length: u64 = 0;
    let mut i = start;

    while i < end {
        if data[i] == b'(' {
            // Find closing parenthesis within the current segment [i+1..end)
            let close_paren_offset = data[i + 1..end]
                .iter()
                .position(|&b| b == b')')
                .ok_or("Unmatched parenthesis within segment")?;
            let close_paren_idx = i + 1 + close_paren_offset; // Index relative to start of data

            let marker_content_bytes = &data[i + 1..close_paren_idx];
            let marker_content_str =
                str::from_utf8(marker_content_bytes).map_err(|_| "Invalid UTF-8 in marker")?;

            let (len_str, rep_str) = marker_content_str
                .split_once('x')
                .ok_or("Invalid marker format: missing 'x'")?;

            let length: usize = len_str.parse().map_err(|_| "Invalid length in marker")?;
            let repetitions: u64 = rep_str
                .parse() // Use u64 directly
                .map_err(|_| "Invalid repetitions in marker")?;

            let sub_start = close_paren_idx + 1;
            let sub_end = sub_start + length;

            // Check bounds: Ensure the referenced segment is within the current segment
            if sub_end > end {
                return Err("Marker references data outside its segment");
            }

            // Recursively calculate the length of the sub-segment
            let sub_length = calculate_decompressed_length_recursive(data, sub_start, sub_end)?;

            decompressed_length += sub_length * repetitions;

            // Move index past the marker and the data block it refers to
            i = sub_end;
        } else {
            decompressed_length += 1;
            i += 1;
        }
    }
    Ok(decompressed_length)
}

fn solve_part2(data: &[u8]) -> Result<u64, &'static str> {
    calculate_decompressed_length_recursive(data, 0, data.len())
}

// --- Main execution ---
fn main() -> Result<(), Box<dyn Error>> {
    let input_path = "../input.txt";
    let input_data = fs::read_to_string(input_path)?;
    let trimmed_data = input_data.trim(); // Remove leading/trailing whitespace
    let bytes = trimmed_data.as_bytes();
    let sep = "=".repeat(20);

    match solve_part1(bytes) {
        Ok(length) => println!("{sep} Part 1 {sep}\nDecompressed length = {}", length),
        Err(e) => eprintln!("Part 1 Error: {}", e),
    }

    match solve_part2(bytes) {
        Ok(length) => println!("{sep} Part 2{sep}\nDecompressed length = {}", length),
        Err(e) => eprintln!("Part 2 Error: {}", e),
    }

    Ok(())
}
